src/lib.rs

```diff
@@ -0,0 +1,111 @@
+    fn record_debug(&mut self, field: &Field, value: &fmt::Debug) {
+        if field.name() == "message" {
+            print!("{:?}: ", value);
+        } else {
+            print!("{} = {:?} ", field.name(), value);
+        }
+    }
+}
+
+pub struct DemoSubscriber {
+    spans: Mutex<Vec<Span>>,
+    stack: Mutex<Vec<u64>>,
+}
+
+impl DemoSubscriber {
+    pub fn new() -> DemoSubscriber {
+        DemoSubscriber {
+            spans: Mutex::default(),
+            stack: Mutex::default(),
+        }
+    }
+}
+
+impl Subscriber for DemoSubscriber {
+    fn enabled(&self, _metadata: &Metadata) -> bool {
+        true
+    }
+
+    fn new_span(&self, attributes: &span::Attributes) -> span::Id {
+        // TODO: correlate spans based on their metadata
+        let mut spans = self.spans.lock().unwrap();
+        let id = spans.len() + 1;
+        spans.push(Span {
+            label: attributes.metadata().name().to_string(),
+            ref_count: 1,
+        });
+        span::Id::from_u64(id as u64)
+    }
+
+    fn record(&self, span: &span::Id, _values: &span::Record) {
+        // Not entirely sure when this is called
+        unimplemented!("Span({}): ", span.into_u64());
+        //values.record(&mut Visitor);
+    }
+
+    fn record_follows_from(&self, span: &span::Id, follows: &span::Id) {
+        unimplemented!(
+            "This is new for me, don't judge: {} - {}",
+            span.into_u64(),
+            follows.into_u64()
+        );
+    }
+
+    fn event(&self, event: &Event) {
+        let level = match *event.metadata().level() {
+            Level::DEBUG => "DEBUG",
+            Level::ERROR => "ERROR",
+            Level::INFO => " INFO", // Whitespace for easy alignment
+            Level::TRACE => "TRACE",
+            Level::WARN => " WARN", // Whitespace for easy alignment
+        };
+        let stack = self.stack.lock().unwrap();
+        let spans = self.spans.lock().unwrap();
+        if let Some(label) = stack.last().map(|id| &spans[(id - 1) as usize].label) {
+            print!("{} - {}: ", level, label);
+        } else {
+            print!("{}: ", level);
+        }
+        event.record(&mut Visitor);
+        println!("(Stack: {:?})", *stack);
+    }
+
+    fn enter(&self, span: &span::Id) {
+        let mut stack = self.stack.lock().unwrap();
+        stack.push(span.into_u64())
+    }
+
+    fn exit(&self, span: &span::Id) {
+        let mut stack = self.stack.lock().unwrap();
+        let top = stack.pop().unwrap();
+        assert_eq!(top, span.into_u64());
+    }
+
+    fn clone_span(&self, id: &span::Id) -> span::Id {
+        let mut spans = self.spans.lock().unwrap();
+        spans[id.into_u64() as usize - 1].ref_count += 1;
+        id.clone()
+    }
+
+    fn drop_span(&self, id: span::Id) {
+        let mut spans = self.spans.lock().unwrap();
+        // This cannot underflow because it's guaranteed to be called correctly
+        spans[id.into_u64() as usize - 1].ref_count -= 1;
```

Comment:
Note that when a span's ref count is zero, the subscriber is free to drop any data stored for that span.

Reply:
True, but here the IDs correspond to vector indices. This makes dropping data difficult, since one can could only free spans at the end of the vector. Reusing indices would help a lot with memory usage.

Reply:
> Reusing indices would help a lot with memory usage.

Yup, this is (again) what `tokio-trace-fmt` does --- you're definitely on the right track! :D

Again this may not matter for a demo, but I wanted to mention it regardless since it's an important part of the API. 

Also, we may care about doing other behaviours when a span's refcount hits zero --- for example, if we're collecting time profiles of how long we spent in a span, the refcount reaching zero is a good time to consider a span as "over" for the purpose of that profile. In `tokio-trace-console`, this is where we would probably want to send a message to the console client that the span has been closed (even if we don't have any internal state to clean up).